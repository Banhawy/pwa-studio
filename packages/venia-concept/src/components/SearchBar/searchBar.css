.root {
    background-color: rgb(var(--venia-grey));
    align-items: center;
    display: flex;
    justify-content: center;
    width: 100%;
    transition-property: opacity, transform, height, padding;
}

.searchBlock {
    composes: root;
    opacity: 0;
    height: 0;
    padding: 0 2rem 0 0.5rem;
    transform: translate3d(0, -100%, 0);
    transition-duration: 192ms;
    transition-timing-function: cubic-bezier(0.4, 0, 1, 1);
    pointer-events: none;
}

.searchBlockOpen {
    composes: root;
    opacity: 1;
    height: 4rem;
    padding: 0 2rem 1rem 0.5rem;
    transform: translate3d(0, 0, 0);
    transition-duration: 224ms;
    transition-timing-function: cubic-bezier(0, 0, 0.2, 1);
    pointer-events: all;
}

.searchBar {
    border: 1px solid rgb(var(--venia-border));
    border-radius: 6px;
    font-size: 1rem;
    height: 3rem;
    padding: 0 2.5rem;
    width: 24rem;
    -webkit-appearance: none;
}

.searchIcon {
    position: relative;
    left: 2.2rem;
    cursor: pointer;
}
