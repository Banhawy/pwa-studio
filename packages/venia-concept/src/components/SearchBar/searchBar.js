import React, { Component } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';

import classify from 'src/classify';
import { executeSearch } from 'src/actions/app';
import defaultClasses from './searchBar.css';

import Icon from 'src/components/Icon';

export class SearchBar extends Component {
    static propTypes = {
        classes: PropTypes.shape({
            root: PropTypes.string,
            searchBlock: PropTypes.string,
            searchBlockOpen: PropTypes.string,
            searchBar: PropTypes.string,
            searchIcon: PropTypes.string
        }),
        isOpen: PropTypes.bool,
        executeSearch: PropTypes.func
    };

    constructor(props) {
        super(props);
        this.searchRef = React.createRef();
        this.state = {isClearIcon: false};
    }

    async componentDidMount() {
        if (this.props.isOpen) {
            this.searchRef.current.focus();
        }
        if (this.props.location.pathname === '/search.html') {
            this.searchRef.current.value = this.props.location.search.substring(7);
            this.setClearIcon(this.searchRef.current.value);
        }
    }

    componentDidUpdate(prevProps) {
        if (this.props.isOpen !== prevProps.isOpen) {
            if (this.props.isOpen == true) {
                this.searchRef.current.focus();
            } else {
                this.searchRef.current.blur();
            }
        }
    }

    enterSearch = event => {
        const searchQuery = this.searchRef.current.value;
        this.setClearIcon(searchQuery);
        if (
            (event.type === 'click' || event.key === 'Enter') &&
            searchQuery !== ''
        ) {
            this.props.executeSearch(searchQuery, this.props.history);
        }
    };

    clearSearch = () => {
        this.searchRef.current.value = '';
        this.searchRef.current.focus();
        this.setClearIcon(this.searchRef.current.value);
    };

    setClearIcon(query) {
        if (query !== '') {
            this.setState(() => {
                return {isClearIcon: true};
            });
        } else {
            this.setState(() => {
                return {isClearIcon: false};
            });
        }
    }

    render() {
        const { classes, isOpen } = this.props;

        const searchClass = isOpen
            ? classes.searchBlockOpen
            : classes.searchBlock;

        const clearIconClass = this.state.isClearIcon ? classes.clearIcon : classes.clearIcon_off;

        return (
            <div className={searchClass}>
                <button
                    className={classes.searchIcon}
                    onClick={this.enterSearch}
                >
                    <Icon name="search" />
                </button>
                <input
                    ref={this.searchRef}
                    className={classes.searchBar}
                    inputMode="search"
                    type="search"
                    placeholder="I'm looking for..."
                    onKeyUp={this.enterSearch}
                />
                <button
                    className={clearIconClass}
                    onClick={this.clearSearch}
                >
                    <Icon name="x" />
                </button>
            </div>
        );
    }
}

export default withRouter(classify(defaultClasses)(SearchBar));
