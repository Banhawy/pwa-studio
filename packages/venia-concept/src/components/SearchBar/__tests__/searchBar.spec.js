import React from 'react';
import { shallow } from 'enzyme';
import { SearchBar } from '../searchBar';

const classes = {
    searchBlockOpen: 'open',
    searchBlock: 'closed'
};

test('When the search bar is expanded, pressing the Enter key will submit.', async () => {
    const mockExecuteSearch = jest.fn();
    let wrapper = shallow(
        <SearchBar
            classes={classes}
            isOpen={true}
            executeSearch={mockExecuteSearch}
        />
    );

    wrapper.instance().searchRef = { current: { value: 'test' } };
    const searchInput = wrapper.find('input');
    searchInput.simulate('focus');
    searchInput.simulate('keyUp', { key: 'Enter' });
    expect(mockExecuteSearch).toHaveBeenCalled();
});

test('When the search bar is minimized, pressing the Enter key will not submit.', async () => {
    const mockExecuteSearch = jest.fn();
    let wrapper = shallow(
        <SearchBar
            classes={classes}
            isOpen={false}
            executeSearch={mockExecuteSearch}
        />
    );

    wrapper.instance().searchRef = { current: { value: 'test' } };
    const searchInput = wrapper.find('input');
    searchInput.simulate('focus');
    searchInput.simulate('keyUp', { key: 'Enter' });
    expect(mockExecuteSearch).toHaveBeenCalledTimes(0);
});

test('When the search icon is clicked, the query in the input component will be submitted.', async () => {
    const mockExecuteSearch = jest.fn();
    let wrapper = shallow(
        <SearchBar
            classes={classes}
            isOpen={true}
            executeSearch={mockExecuteSearch}
        />
    );

    wrapper.instance().searchRef = { current: { value: 'test' } };
    const searchButton = wrapper.find('button').at(0);
    searchButton.simulate('focus');
    searchButton.simulate('click', { type: 'click' }, { button: '0' });
    expect(mockExecuteSearch).toHaveBeenCalled();
});

test('When the input component is empty, pressing the enter key will not search.', async () => {
    const mockExecuteSearch = jest.fn();
    let wrapper = shallow(
        <SearchBar
            classes={classes}
            isOpen={true}
            executeSearch={mockExecuteSearch}
        />
    );

    wrapper.instance().searchRef = { current: { value: '' } };
    const searchInput = wrapper.find('input');
    searchInput.simulate('focus');
    searchInput.simulate('keyUp', { key: 'Enter' });
    expect(mockExecuteSearch).toHaveBeenCalledTimes(0);
});

test('When the clear button is pressed, any text in the input component is removed.', async () => {
    const mockFocus = jest.fn();
    let wrapper = shallow(<SearchBar classes={classes} isOpen={true} />);

    const searchInput = wrapper.find('input');
    const clearButton = wrapper.find('button').at(1);
    wrapper.instance().searchRef = {
        current: { value: 'test', focus: mockFocus }
    };
    searchInput.simulate('change');
    clearButton.simulate('click');
    expect(wrapper.instance().searchRef.current.value).toBe('');
});

test('When the input component is empty, the clear button is not displayed.', async () => {
    let wrapper = shallow(<SearchBar classes={classes} isOpen={true} />);

    wrapper.instance().searchRef = { current: { value: '' } };

    expect(wrapper.state('showClearIcon')).toBe(false);
});

test('When the input element has text, the clear button is displayed.', async () => {
    const mockEvent = { type: 'keyUp' };
    let wrapper = shallow(<SearchBar classes={classes} isOpen={true} />);

    wrapper.instance().searchRef = { current: { value: 'test' } };
    // Call the onKeyUp event handler directly since we can't alter the ref.
    wrapper.instance().enterSearch(mockEvent);

    expect(wrapper.state('showClearIcon')).toBe(true);
});
