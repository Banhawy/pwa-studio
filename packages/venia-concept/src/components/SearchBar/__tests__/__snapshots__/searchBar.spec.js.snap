// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div>
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" class=\\"feather feather-search\\"><circle cx=\\"11\\" cy=\\"11\\" r=\\"8\\"></circle><line x1=\\"21\\" y1=\\"21\\" x2=\\"16.65\\" y2=\\"16.65\\"></line></svg>",
        }
      }
    />
    <input
      name="search_query"
      onBlur={[Function]}
      onChange={[Function]}
      value="test"
    />
    <p />
    <button
      onClick={[Function]}
      type="button"
    >
      <span>
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" class=\\"feather feather-x\\"><line x1=\\"18\\" y1=\\"6\\" x2=\\"6\\" y2=\\"18\\"></line><line x1=\\"6\\" y1=\\"6\\" x2=\\"18\\" y2=\\"18\\"></line></svg>",
            }
          }
        />
      </span>
    </button>
  </div>
</form>
`;
